### ISSUE TITLE
Cannot access the query object returned from an async function

### ISSUE THREAD

**Issue by ivosabev** [tgriesser/knex#2940]
I assume this is due to lack of understanding of the internals of Knex, but it is very common to have a async/await function that awaits on some other async call and then proceed to generate a Knex query which I would like to return without executing so I can continue to augment it.

Here is a very naive example that is currently impossible:
```
async function async getPostsForUser(userId) {
	const user = await this.getUser(userId);
  	return db.select('users').where('userId', user.id);
}

const q = (await getPostsForUser()).limit(10); // would expect Knex query

const result = await q;
```

**Comment by wubzz** [comment]
```javascript
const q = getPostsForUser().limit(10); // queryBuilder -> call limit(10)

const result = await q; //select * from users where "userId" = ? LIMIT 10
```

**Comment by wubzz** [comment]
Since querybuilder is a 'thenable' calling await on it will issue the query and return the result. The example is a bit confusing to me. The function takes userId but also runs a lookup from this userId -> user.id. Seems redundant?

There are other ways of solving this but it does not involve using await at all, which I'm guessing is what you really want.

**Comment by elhigu** [comment]
Not knex problem. As @wubzz mentioned, that is just the way how promises / async / await works with thenables. If you want to return `thenable` from async function without causing it to resolve, you have to wrap it to for example inside an object for returning.

**Comment by ivosabev** [comment]
@elhigu Thanks for your explanation, I understand that is not per se Knex issue, but could you provide an example of how it could be done with the object wrapping?

**Comment by elhigu** [comment]
```
async function async getPostsForUser(userId) {
	const user = await this.getUser(userId);
  	return { builder: db.select('users').where('userId', user.id) };
}

const q = (await getPostsForUser()).builder.limit(10);

const result = await q;
```

**Comment by ivosabev** [comment]
I expected promises to be chained and resolved one by one by every single await. Thanks for the example, maybe it should be included somewhere as a FAQ.

**Comment by whyayala** [comment]
I also feel this should be documented. Had the same question myself which led me here.

**Comment by JenChampagne** [comment]
Same, I tormented myself for way too long before checking here.

**Comment by JenChampagne** [comment]
Agreed. Though in my situation, I had something like this with a BlueBird Promise that was kicked off by another different method:

```
otherMethod()
.then(values => knex.select(table).where('somefield', values.x))
.then(knex => expectingBuilderHereButGotResultsInstead);
```

I understand why it works this way, but it would be nice to have a method to chain on that will prevent the knex query builder from executing in a context like this.

**Comment by elhigu** [comment]
>  it would be nice to have a method to chain on that will prevent the knex query builder from executing in a context like this.

Then it would be nice feature to have on every library that implements `thenable` pattern from promise spec. I don't think that is a good idea, since all JS developers should know how promises work and know their way to go around it.

I also learned that in some code that I was writing years ago, which was not related to knex in any way... Also arrow functions is fine way to go around the problem, but to me wrapping to object seems more clear.

**Comment by ivosabev** [comment]
Actually @JenChampagne idea is not that bad and it could look similar to the `stream()` function (https://knexjs.org/#Interfaces-Streams).

```
var stream = knex.select('*').from('users').stream();
var query = knex.select('*').from('users').query();
```

Then we can have:
```
async function getUsers() {
  const organization = await getViewerOrganization();
  const query = knex.select('*').from('users').where('organization_id', organization.id).query();
  return query;
}

```

**Comment by elhigu** [comment]
@ivosabev that wouldn't work. Returned `query` would not be `thenable` anymore and for example this wouldn't do anything:

```
async function getUsers() {
  const organization = await getViewerOrganization();
  const query = knex
    .select('*')
    .from('users')
    .where('organization_id', organization.id)
    .query();

  return query;
}

query = await getUsers();
let users = await query;
```

**Comment by elhigu** [comment]
>  I just want a way to force it to not be automatically promisified.

By promise spec, that would require removing or disabling `.then` method from query builder. That would mean that after returning that query builder from `async` method one would have to somehow enable, `.then` method again.. so your code would start to look something like this.

```
async foo() {
  return knex('foo').disableThen();
}

query = await foo();
await query.enableThen();
```

Unless disable method would be automatically set to re-enable then after it has been resolved for the first time. In that case if there would be some nested hierarchy where it is passed through multiple async funcs, you would need to disable it on every step... which just sounds too magical and hacky and would cause even more confusion. The way to get around this problem by wrapping the value is more clear solution to ever JS developer, whom are not familiar with all peculiarities of knex.

> It has nothing to do with experience levels of JavaScript.

I believe that deep understanding how promises work has exactly something to do with that 🤔

**Comment by JenChampagne** [comment]
> > I just want a way to force it to not be automatically promisified.
> 
> By promise spec, that would require removing or disabling `.then` method from query builder. That would mean that after returning that query builder from `async` method one would have to somehow enable, `.then` method again.. so your code would start to look something like this.
> 
> ```
> async foo() {
>   return knex('foo').disableThen();
> }
> 
> query = await foo();
> let allTheFoos = await query.enableThen();
> ```

That is exactly what I mean by having a way to force it to not be promisified. Though I would be a little more descriptive with the method names. Going back to my earlier example:

```
otherMethod()
.then(values => knex.select(table).where('somefield', values.x).partialQuery())
.then(knex => expectingBuilderHereButGotResultsInstead.orderBy('otherfield', 'desc').execute())
.then(results => queryResultsHere);
```

> Unless disable method would be automatically set to re-enable then after it has been resolved for the first time. In that case if there would be some nested hierarchy where it is passed through multiple async funcs, you would need to disable it on every step... which just sounds too magical and hacky and would cause even more confusion. The way to get around this problem by wrapping the value is more clear solution to ever JS developer, whom are not familiar with all peculiarities of knex.

I don't think any of that is necessary. It's not a wildly uncommon feature to call an "execute" type of function to cause an sql query to be sent to the database for processing.

> > It has nothing to do with experience levels of JavaScript.
> 
> I believe that deep understanding how promises work has exactly something to do with that 🤔

If I didn't understand how Promises worked, I wouldn't have requested the above feature as I stated above:
> I understand why it works this way, but it would be nice to have a method to chain on that will prevent the knex query builder from executing in a context like this.

**Comment by ivosabev** [comment]
Having `builder()` or `query()` and `execute()` function sounds reasonable and in my opinion it could live in parallel with the current design that returns thenable.

**Comment by elhigu** [comment]
It could live in parallel, but I'm heavily against it, because of various reasons (mostly that it will cause more harm than good for `knex` project which is already too bloated with various APIs it offers):

- it would be more confusing to have multiple types of query builders going around, which some are `thenables` and others are not
- that additional API is not much better than,  just wrapping normal builder
- it will cause more maintenance burden, when all knex parts need to make sure to work with `non-thenable` query builder, 
- there will be more related bug reports when there are confusion about it or when people find all corner cases where those query builders are not working properly
- tl;dr this kind of seemingly small features which are easy to workaround outside of knex  very easily add up to become big burden in the long run

If knex will have proper support for extending it with plugins, that would be good place to add this kind of less commonly needed APIs as separate packages. For now I suggest you just add your own wrapper method for passing query builder instance without executing them from async functions. Like:

```
function partialQuery(bulider) {
  return { query: builder };
}

otherMethod()
.then(values => partialQuery(knex.select(table).where('somefield', values.x)))
.then(knexBuilder => knexBuilder.query.orderBy('otherfield', 'desc'))
.then(results => queryResultsHere);
```

I don't think I have anything else to add to this issue, those provided workarounds should be enough for anyone.
